dbms/src/Flash/Executor/PipelineExecutorStatus.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <assert.h>
+
+namespace DB
+{
+ExecutionResult PipelineExecutorStatus::toExecutionResult()
+{
+    auto get_err_msg = getErrMsg();
+    return get_err_msg.empty()
+        ? ExecutionResult::success()
+        : ExecutionResult::fail(get_err_msg);
+}
+
+String PipelineExecutorStatus::getErrMsg()
+{
+    std::lock_guard lock(mu);
+    return err_msg;
+}
+
+void PipelineExecutorStatus::toError(String && err_msg_)
+{
+    {
+        std::lock_guard lock(mu);
+        if (!err_msg.empty())
+            return;
+        err_msg = err_msg_.empty() ? empty_err_msg : std::move(err_msg_);
+    }
+    cancel();
```

Comment:
Implicitly cancelling in this method is not a good semantic. Better explicitly `toError` then `cancel` all in caller.

Also, do we really tolerate calling `toError()` multiple times and silently discards the second and subsequent calls?

Reply:
renamed `toError` to `onErrorOccurred`. And doing multiple err-related things in `onErrorOccurred` is consistent with the semantics.